File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/function/system/date/EpochToDate.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.function.reactive.AbstractReactiveFunction;
import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.model.EventResult;
import com.fincity.nocode.kirun.engine.model.FunctionOutput;
import com.fincity.nocode.kirun.engine.model.FunctionSignature;
import com.fincity.nocode.kirun.engine.model.Parameter;
import com.fincity.nocode.kirun.engine.namespaces.Namespaces;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonPrimitive;

import reactor.core.publisher.Mono;

public class EpochToDate extends AbstractReactiveFunction {

	private static final String EPOCH = "epoch";

	private static final String OUTPUT = "date";

	private static final String ERROR_MSG = "Please provide a valid value for epoch.";

	@Override
	public FunctionSignature getSignature() {

		return new FunctionSignature().setNamespace(Namespaces.DATE)
		        .setName("EpochToDate")
		        .setParameters(Map.of(EPOCH, Parameter.of(EPOCH, new Schema()
		                .setAnyOf(List.of(Schema.ofInteger(EPOCH), Schema.ofLong(EPOCH), Schema.ofString(EPOCH))))));

	}

	@Override
	protected Mono<FunctionOutput> internalExecute(ReactiveFunctionExecutionParameters context) {

		var epochIp = context.getArguments()
		        .get(EPOCH);

		if (!epochIp.isJsonPrimitive())
			throw new KIRuntimeException(ERROR_MSG);

		JsonPrimitive epochPrimitive = epochIp.getAsJsonPrimitive();

		if (epochPrimitive.isBoolean())
			throw new KIRuntimeException(ERROR_MSG);

		Long longDate = epochPrimitive.isNumber() ? epochPrimitive.getAsLong()
		        : Long.parseLong(epochPrimitive.getAsString());

		Long updatedValue = longDate > 999999999999L ? longDate : longDate * 1000;

		ZonedDateTime zdt = ZonedDateTime.ofInstant(Instant.ofEpochMilli(updatedValue), ZoneId.of("UTC"));

		DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");

		return Mono.just(
		        new FunctionOutput(List.of(EventResult.outputOf(Map.of(OUTPUT, new JsonPrimitive(dtf.format(zdt)))))));
	}

}

File: kirun-js/src/engine/function/system/date/EpochToDate.ts
import { Schema } from "../../../json/schema/Schema";
import { FunctionOutput } from "../../../model/FunctionOutput";
import { FunctionSignature } from "../../../model/FunctionSignature";
import { Event } from '../../../model/Event';
import { Parameter } from "../../../model/Parameter";
import { Namespaces } from "../../../namespaces/Namespaces";
import { FunctionExecutionParameters } from "../../../runtime/FunctionExecutionParameters";
import { AbstractFunction } from "../../AbstractFunction";
import { isNullValue } from "../../../util/NullCheck";
import { KIRuntimeException } from "../../../exception/KIRuntimeException";
import { EventResult } from "../../../model/EventResult";


const VALUE = 'epoch';
const OUTPUT = 'date';
const  ERROR_MSG: string = "Please provide a valid value for epoch.";

const SIGNATURE = new FunctionSignature('EpochToDate')
    .setNamespace(Namespaces.DATE)
    .setParameters(
        new Map([
            [
                VALUE,
                new Parameter(
                    VALUE,
                    new Schema().setAnyOf([
                        Schema.ofInteger(VALUE),
                        Schema.ofLong(VALUE),
                        Schema.ofString(VALUE),
                    ]),
                ),
            ],
        ]),
    )
    .setEvents(
        new Map([
            Event.outputEventMapEntry(
                new Map([[OUTPUT, Schema.ofRef(`${Namespaces.DATE}.timeStamp`)]]),
            ),
        ]),
    );

export class EpochToDate extends AbstractFunction{

    public getSignature(): FunctionSignature {
      return SIGNATURE;
    }
    
    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {

       var epoch : any =  context.getArguments()?.get(VALUE);

       if(isNullValue(epoch))
        throw new KIRuntimeException(ERROR_MSG);

       if(typeof epoch === 'boolean')
            throw new KIRuntimeException(ERROR_MSG);

       if(typeof epoch === 'string')
            epoch  = parseInt(epoch)

       if(isNaN(epoch))
        throw new KIRuntimeException(ERROR_MSG);

       epoch = epoch > 999999999999 ? epoch : epoch * 1000;

      
       return new FunctionOutput([
           EventResult.outputOf(new Map([
            [OUTPUT, new Date(epoch).toISOString()]
            ]))
       ]);

    }
    
}